src/Pos/Communication/Limits.hs

```diff
@@ -6,134 +6,178 @@
+instance MessageLimitedPure PublicKey where
+    msgLenLimit = 64
+
+instance MessageLimitedPure a => MessageLimitedPure (AsBinary a) where
+    msgLenLimit = coerce (msgLenLimit :: Limit a) + 20
 
-mcCommitmentMsgLenLimit :: Limit GtMsgContents
-mcCommitmentMsgLenLimit = MCCommitment <$> msgLenLimit
+instance MessageLimitedPure SecretProof where
+    msgLenLimit = 64
+
+instance MessageLimitedPure VssPublicKey where
+    msgLenLimit = 33
+
+instance MessageLimitedPure EncShare where
+    msgLenLimit = 101
+
+instance MessageLimitedPure Share where
+    msgLenLimit = 101 --4+33+64
+
+instance MessageLimitedPure PVSS.Commitment where
+    msgLenLimit = 33
 
-mcOpeningLenLimit :: Limit GtMsgContents
-mcOpeningLenLimit = 62
+instance MessageLimitedPure PVSS.ExtraGen where
+    msgLenLimit = 33
 
-mcSharesMsgLenLimit :: Limit GtMsgContents
-mcSharesMsgLenLimit =
-    MCShares <$> msgLenLimit <+> multiMap commitmentsNumLimit
+instance MessageLimitedPure (AbstractHash Blake2b_224 a) where
+    msgLenLimit = 28
 
-mcVssCertificateLenLimit :: Limit GtMsgContents
-mcVssCertificateLenLimit = 169
+instance MessageLimitedPure (AbstractHash Blake2b_256 a) where
+    msgLenLimit = 32
 
 ----------------------------------------------------------------------------
--- Instances
+---- GodTossing
 ----------------------------------------------------------------------------
 
-instance MessageLimited (MsgBlock ssc) where
-    type LimitType (MsgBlock ssc) = Limit (MsgBlock ssc)
-    getMsgLenLimit _ = Limit <$> DB.getMaxBlockSize
+-- | Upper bound on number of `PVSS.Commitment`s in single
+-- `Commitment`.  Actually it's a maximum number of participants in
+-- GodTossing. So it also limits number of shares, for instance.
+commitmentsNumLimit :: DB.MonadGStateCore m => m Int
+commitmentsNumLimit =
+    -- succ is just in case
+    succ . ceiling . recip . coinPortionToDouble . bvdMpcThd <$>
+    DB.gsAdoptedBVData
 
-instance MessageLimited MsgGetHeaders where
-    type LimitType MsgGetHeaders = Limit MsgGetHeaders
-    getMsgLenLimit _ = return $
-        MsgGetHeaders <$> vector maxGetHeadersNum <+> msgLenLimit
-      where
-        maxGetHeadersNum = ceiling $
-            log ((fromIntegral :: Int -> Double) Const.blkSecurityParam) + 5
+instance MessageLimited SecretSharingExtra where
+    getMsgLenLimit _ = do
+        numLimit <- commitmentsNumLimit
+        return $ SecretSharingExtra <$> msgLenLimit <+> vector numLimit
 
-instance MessageLimited (MsgHeaders ssc) where
-    type LimitType (MsgHeaders ssc) = Limit (MsgHeaders ssc)
+instance MessageLimited (AsBinary SecretSharingExtra) where
+    -- 20 is a magic constant copy-pasted from 'MessageLimitedPure' instance
+    getMsgLenLimit _ =
+        coerce . (20 +) <$> getMsgLenLimit (Proxy @SecretSharingExtra)
+
+instance MessageLimited Commitment where
     getMsgLenLimit _ = do
-        headerLimit <- Limit <$> DB.getMaxHeaderSize
+        extraLimit <- getMsgLenLimit (Proxy @(AsBinary SecretSharingExtra))
```

Comment:
No, it can't, because type families are not injective. There can be more than one type for which `TypeLimit` is `Limit (AsBinary SecretSharingExtra)`.